File: hw02/fourbuttons.js
#!/usr/bin/env node
/* File: fourbuttons.js
 * Author: Andrew Mueller
 *
 * This file takes fur buttons and toggles four LEDs, One for each button.
 * The buttons are plugged in to GP0 while the LEDs are plugged in through GP! 
 * making use of  220 ohm resistors.
*/

//Setting up required variables
var b = require('bonescript');

var led = ["GP1_3", "GP1_4", "RED", "GREEN"];
var state = [0, 0, 0 ,0];
var prev = [1, 1, 1, 1];
var button = ["GP0_3", "GP0_4", "GP0_5", "GP0_6"];

//set pinMode
b.pinMode(led[0], b.OUTPUT);
b.pinMode(led[1], b.OUTPUT);
b.pinMode(led[2], b.OUTPUT);
b.pinMode(led[3], b.OUTPUT);

b.pinMode(button[0], b.INPUT);
b.pinMode(button[1], b.INPUT);
b.pinMode(button[2], b.INPUT);
b.pinMode(button[3], b.INPUT);

//Attaching interrups
b.attachInterrupt(button[0], toggleB0, b.CHANGE);  //Could potetally be changesto falling edge but that would mess with my debounce method
b.attachInterrupt(button[1], toggleB1, b.CHANGE);
b.attachInterrupt(button[2], toggleB2, b.CHANGE);
b.attachInterrupt(button[3], toggleB3, b.CHANGE);

//Init the leds (on)
b.digitalWrite(led[0], 0);
b.digitalWrite(led[1], 0);
b.digitalWrite(led[2], 0);
b.digitalWrite(led[3], 0);

//The interups to be used by the buttons
function toggleB0(x) {
  setTimeout(function(){ // debounces the button
    if (x.value == 0 && prev[0] == 1) { //checks for falling edge
      (state[0] == 0) ? state[0] = 1 : state[0] = 0;  //toggles the state
      b.digitalWrite(led[0], state[0]);  //changes the light
    }
  prev[0] = x.value;  // updates the previous value
  }, 25); 
}

function toggleB1(x) {
  setTimeout(function(){
    if (x.value == 0 && prev[1] == 1) {
      (state[1] == 0) ? state[1] = 1 : state[1] = 0;
      b.digitalWrite(led[1], state[1]);
    }
  prev[1] = x.value;
  }, 25); 
}

function toggleB2(x) {
  setTimeout(function(){
    if (x.value == 0 && prev[2] == 1) {
      (state[2] == 0) ? state[2] = 1 : state[2] = 0;
      b.digitalWrite(led[2], state[2]);
    }
  prev[2] = x.value;
  }, 25); 
}

function toggleB3(x) {
  setTimeout(function(){
    if (x.value == 0 && prev[3] == 1) {
      (state[3] == 0) ? state[3] = 1 : state[3] = 0;
      b.digitalWrite(led[3], state[3]);
    }
  prev[3]= x.value;
  }, 25); 
}

File: hw02/fourbuttons.py
#!/usr/bin/env python3
# File: fourbuttons.py
# Author: Andrew Mueller
#
# This file takes input form 4 buttons on GP0 and toggles 4 leds on GP1

# Import PyBBIO library:
import Adafruit_BBIO.GPIO as GPIO
import time

button = ["GP0_3", "GP0_4", "GP0_5", "GP0_6"]  # the buttons used
led = ["GP1_3", "GP1_4", "RED", "GREEN"] # LEDs used

# Set the GPIO pins:
GPIO.setup(led[0],    GPIO.OUT)
GPIO.setup(led[1],    GPIO.OUT)
GPIO.setup(led[2],    GPIO.OUT)
GPIO.setup(led[3],    GPIO.OUT)

GPIO.setup(button[0], GPIO.IN)
GPIO.setup(button[1], GPIO.IN)
GPIO.setup(button[2], GPIO.IN)
GPIO.setup(button[3], GPIO.IN)

# Turn on the LEDs
GPIO.output(led[0], 1)
GPIO.output(led[1], 1)
GPIO.output(led[2], 1)
GPIO.output(led[3], 1)

# Map buttons to LEDs
map = {button[0]: led[0], button[1]: led[1], button[2]: led[2], button[3]: led[3]}

def updateLED(channel):      #interrupt
    state = GPIO.input(channel)     #get the state
    GPIO.output(map[channel], state) #toggle the LED

print("Running...")

GPIO.add_event_detect(button[0], GPIO.BOTH, callback=updateLED) #adding the interrupt
GPIO.add_event_detect(button[1], GPIO.BOTH, callback=updateLED)
GPIO.add_event_detect(button[2], GPIO.BOTH, callback=updateLED)
GPIO.add_event_detect(button[3], GPIO.BOTH, callback=updateLED)

try:
    while True:
        time.sleep(100)   # Let other processes run

except KeyboardInterrupt:
    print("Cleaning Up")
    GPIO.cleanup()
